faststreams/input_stream.nim

```diff
@@ -1,192 +1,298 @@
+  ## effects analysis of Nim - reading from global let variables
+  ## is considered a side-effect.
+  {.noSideEffect.}:
+    unsafeAddr vtable
+
+proc fileInput*(filename: string): InputStreamHandle =
+  let
+    memFile = memfiles.open(filename)
+    head = cast[ptr byte](memFile.mem)
+    fileSize = memFile.size
+
+  var stream = FileInputStream(
+    vtable: vtableAddr FileStreamVTable,
+    head: head,
+    bufferEnd: offset(head, fileSize),
+    bufferEndPos: fileSize,
+    file: memFile)
 
-proc openFile*(filename: string): ByteStreamVar =
-  new result
-  var memFile = memfiles.open(filename)
-  result.head = cast[ptr byte](memFile.mem)
   when debugHelpers:
-    result.bufferStart = result.head
-  result.bufferEnd = offset(result.head, memFile.size)
-  result.bufferEndPos = memFile.size
-  result.closeStream = proc = close memFile
-
-proc init*(T: type ByteStream,
-           mem: openarray[byte],
-           reader = StreamReader(nil),
-           asyncReader = AsyncStreamReader(nil)): ByteStream =
-  # TODO: the result should use `from mem` once it's supported
-  result.head = unsafeAddr mem[0]
-  result.bufferEnd = offset(result.head, mem.len)
-  result.bufferEndPos = mem.len
-  result.reader = reader
-  result.asyncReader = asyncReader
-
-proc memoryStream*(mem: openarray[byte]): ByteStreamVar = # TODO: mark the result with `from mem`
-  new result
-  result[] = ByteStream.init(mem)
-
-proc memoryStream*(str: string): ByteStreamVar = # TODO: mark the result with `from str`
-  new result
-  result[] = init(ByteStream, str.toOpenArrayByte(0, str.len - 1))
-
-proc init*(T: type BufferedStream,
-           reader = StreamReader(nil),
-           asyncReader = AsyncStreamReader(nil)): BufferedStream =
-  result.init result.buffer, reader, asyncReader
-
-proc endPos*(s: InputStreamVar): int =
-  # TODO This needs to use a VTable for `system.File` based streams
+    stream.bufferStart = head
+
+  InputStreamHandle(s: stream)
+
+proc memoryInput*(mem: openarray[byte]): InputStreamHandle =
+  let head = unsafeAddr mem[0]
+  InputStreamHandle(s: InputStream(
+    head: head,
+    bufferEnd: offset(head, mem.len),
+    bufferEndPos: mem.len))
+
+proc memoryInput*(str: string): InputStreamHandle =
+  memoryInput str.toOpenArrayByte(0, str.len - 1)
+
+# TODO: Is this used, should we deprecate it?
+proc endPos*(s: InputStream): int =
+  doAssert s.vtable == nil or s.vtable.getLenSync != nil
   return s.bufferEndPos
 
-proc syncRead(s: var ByteStream): bool =
-  let bytesRead = s.reader(s.bufferStart, s.bufferSize)
+# TODO The return type here could be Option[Natural] if Nim had
+# the Option[range] optimisation that will make it equvalent to `int`.
+proc len*(s: InputStream): int {.raises: [Defect, IOError].} =
+  if s.vtable == nil:
+    distance(s.head, s.bufferEnd)
+  elif s.vtable.getLenSync != nil:
+    s.vtable.getLenSync(s)
+  else:
+    lengthUnknown
+
+template len*(s: AsyncInputStream): int =
```

Comment:
Can you give me an example of an async stream with a known length?

Reply:
* reading a file from a networked filesystem
* reading a http page with known size
* reading a length-prefixed stream where the first bytes are needed to determine the length

Reply:
In the network file example, I think the stream creation should be async, but it will result in a stream with a length stored in a local variable (so it can be obtained in a sync way).

The other examples are not quite appropriate, because they are protocol-specific and the length is part of the stream content, it's not metadata associated with the stream. Thus, the length handling code will be part of the protocol implementation that consumes data from the stream.

You need to describe a situation where the length can be considered metadata for the content and it makes sense to obtain this metadata in a lazy fashion.